docs/signaling_architecture.md

```diff
@@ -0,0 +1,97 @@
+3) **JMS** responds positively.
+4) **BE** responds to client with generated token.
+5) **CL** opens WebSocket connection to **JMS**, flow of signaling messages begins.
+
+**JMS** can diferentiate between clients by the opened WebSocket connection so it knows who is the sender of
+incoming Media Events and where to send generated Media Events.
+
+**Advantages:**
+
+- Easier option, the user doesn't have to do much more than passing some secret (authentication token) to the **CL**, client SDK handles the rest.
+- Both **CL** and **JMS** can tell when signaling connection was broken. That prevents situation when signaling connection between **CL** and **BE** was broken, but **BE**
+did not notify **JMS** of such occurence, is that case media is still flowing with no signaling (when new peer joins, tracks are not renegotiated, etc.).
+
+**Disadvantages:**
+
+- Client has to participate in the authentication process (so they probably need to be authenticated by **BE** and then authenticated again to open connection to **JMS**).
+- Some events can get desynchronized. Imagine scenerio where **JMS** sent notification to **BE** that recording has begun (it is **BE**'s responibility
+to propagate this information to peers, so they can e.g. show some icon indicator). At the same time, media and signaling connections broke for some reason,
+so for a brief moment (until **JMS** passes that information to **BE** and **BE** passes it to **CL**) **CL** might think that their screen is being recorded,
+even though that is not the case (situation in this example can be prevented, but I hope you get the gist).
+
+### Using **BE** as a middleman
+
+Client SDK generates signaling messages, but requires the user to implement logic that handles forwarding them to **JMS** (using **BE** in described scenario).
+
+**Exmaple:**
+
+1) **CL** sends request to **BE** to join some room (meeting, webinar).
+2) **BE** sends request to **JMS** (*add_peer*).
+3) **JMS** responds positively.
+4) **BE** responds to client positively.
+5) **CL** starts generating signaling messages which are forwarded to **JMS** by **BE**.
+
+In that case we need a way to identify the sender of signaling messages, we thought of 2 approaches (you'r welcome to suggest a better one):
+
+- make **BE** open WebSocket connection to **JMS** for every client (here seems like a very bad idea, makes a bit more sense in the mixed approach described later),
+- tag every signaling message (Media Event) with e.g. `peer_id`, requires only one WebSocket connection between **BE** and **JMS** (should SDK do this, or should it be the users responsibility?).
+
+**Advantages:**
+
+- Easy to implement (from our, Jellyfish developers, perspective).
+- **CL** does not require additional connection and authentication, everything is handled by **BE**.
+
+**Disadvantages:**
+
+- Harder to implement by the user, much more error prone (isn't that the point of Jellyfish to make it as simple as possible?).
```

Comment:
I think we were also aiming to minimize boilerplate, and in this approach, you would basically have to implement the signaling channel yourself, which generates boilerplate code.